File: async/src/main/java/info/xiaomo/anysc/controller/TestController.java
package info.xiaomo.anysc.controller;

import info.xiaomo.anysc.task.AsyncTask;
import info.xiaomo.core.base.BaseController;
import info.xiaomo.core.base.Result;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.concurrent.Future;

/**
 * 把今天最好的表现当作明天最新的起点．．～
 * いま 最高の表現 として 明日最新の始発．．～
 * Today the best performance  as tomorrow newest starter!
 * Created by IntelliJ IDEA.
 *
 * @author : xiaomo
 * github: https://github.com/xiaomoinfo
 * email: xiaomo@xiaomo.info
 * <p>
 * Date: 2016/11/15 15:12
 * Description: 用户实体类
 * Copyright(©) 2015 by xiaomo.
 **/

@RestController
@RequestMapping("/")
public class TestController extends BaseController {

    private final AsyncTask task;

    @Autowired
    public TestController(AsyncTask task) {
        this.task = task;
    }

    @RequestMapping(value = "/", method = RequestMethod.GET)
    public Result task() throws Exception {
        long start = System.currentTimeMillis();

        Future<String> task1 = task.doTaskOne();
        Future<String> task2 = task.doTaskTwo();
        Future<String> task3 = task.doTaskThree();

        while (true) {
            if (task1.isDone() && task2.isDone() && task3.isDone()) {
                // 三个任务都调用完成，退出循环等待
                break;
            }
            Thread.sleep(1000);
        }

        long end = System.currentTimeMillis();

        System.out.println("任务全部完成，总耗时：" + (end - start) + "毫秒");
        return new Result<>(end - start);
    }


    /**
     * 查找所有(不带分页)
     *
     * @return result
     */
    @Override
    public Result<List> findAll() {
        return null;
    }

    /**
     * 带分页
     *
     * @param start    起始页
     * @param pageSize 页码数
     * @return result
     */
    @Override
    public Result<Page> findAll(@PathVariable int start, @PathVariable int pageSize) {
        return null;
    }

    /**
     * 根据id查看模型
     *
     * @param id id
     * @return result
     */
    @Override
    public Result findById(@PathVariable Long id) {
        return null;
    }

    /**
     * 根据名字查找模型
     *
     * @param name name
     * @return result
     */
    @Override
    public Result findByName(@PathVariable String name) {
        return null;
    }

    /**
     * 根据名字删除模型
     *
     * @param name name
     * @return result
     */
    @Override
    public Result<Boolean> delByName(@PathVariable String name) {
        return null;
    }

    /**
     * 根据id删除模型
     *
     * @param id id
     * @return result
     */
    @Override
    public Result<Boolean> delById(@PathVariable Long id) {
        return null;
    }

    /**
     * 添加模型
     *
     * @param model model
     * @return result
     */
    @Override
    public Result<Boolean> add(@RequestBody Object model) {
        return null;
    }

    /**
     * 更新
     *
     * @param model model
     * @return result
     */
    @Override
    public Result<Boolean> update(@RequestBody Object model) {
        return null;
    }

    /**
     * 批量删除
     *
     * @param ids ids
     * @return result
     */
    @Override
    public Result<Boolean> delByIds(@PathVariable List ids) {
        return null;
    }
}

File: async/src/main/java/info/xiaomo/anysc/controller/TestController.kt
package info.xiaomo.anysc.controller

import info.xiaomo.anysc.task.AsyncTask
import info.xiaomo.core.base.Result
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.RestController

/**
 * 把今天最好的表现当作明天最新的起点．．～
 * いま 最高の表現 として 明日最新の始発．．～
 * Today the best performance  as tomorrow newest starter!
 * Created by IntelliJ IDEA.
 *
 * @author : xiaomo
 * github: https://github.com/xiaomoinfo
 * email: xiaomo@xiaomo.info
 *
 *
 * Date: 2016/11/15 15:12
 * Description: 用户实体类
 * Copyright(©) 2015 by xiaomo.
 */

@RestController
@RequestMapping("/")
class TestController @Autowired
constructor(private val task: AsyncTask) {

    @RequestMapping(value = "/", method = arrayOf(RequestMethod.GET))
    @Throws(Exception::class)
    fun task(): Result<*> {
        val start = System.currentTimeMillis()

        val task1 = task.doTaskOne()
        val task2 = task.doTaskTwo()
        val task3 = task.doTaskThree()

        while (true) {
            if (task1.isDone && task2.isDone && task3.isDone) {
                // 三个任务都调用完成，退出循环等待
                break
            }
            Thread.sleep(1000)
        }

        val end = System.currentTimeMillis()

        println("任务全部完成，总耗时：" + (end - start) + "毫秒")
        return Result(end - start)
    }

}
